cmd/common/oci.go

```diff
@@ -0,0 +1,140 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends/console"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+func NewRunCommand(rootCmd *cobra.Command, runtime runtime.Runtime, hiddenColumnTags []string) *cobra.Command {
+	runtimeGlobalParams := runtime.GlobalParamDescs().ToParams()
+	runtime.Init(runtimeGlobalParams)
+
+	// Add operator global flags
+	operatorsGlobalParamsCollection := operators.GlobalParamsCollection()
+
+	cmd := &cobra.Command{
+		Use:          "oci",
+		Short:        "run image based gadget",
+		SilenceUsage: true, // do not print usage when there is an error
+		// We have to disable flag parsing in here to be able to handle certain
+		// flags more dynamically and have `--help` also react to those changes.
+		// Instead, we need to manually
+		// * call cmd.ParseFlags()
+		// * handle `--help` after changing the params dynamically
+		// * handle everything that could have been handled inside
+		//   `PersistentPreRun(E)` of a parent cmd, as the flags wouldn't have
+		//   been parsed there (e.g. --verbose)
+		DisableFlagParsing: true,
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			err := runtime.Init(runtimeGlobalParams)
+			if err != nil {
+				return fmt.Errorf("initializing runtime: %w", err)
+			}
+			defer runtime.Close()
+
+			// we need to re-enable flag parsing, as utils.ParseEarlyFlags() would
+			// not do anything otherwise
+			cmd.DisableFlagParsing = false
+
+			// Parse flags that are known at this time, like the ones we get from the gadget descriptor
+			if err := utils.ParseEarlyFlags(cmd, args); err != nil {
+				return err
+			}
+
+			// gadget parameters that are only available after contacting the server
+			gadgetParams := make(params.Params, 0)
+
+			// Before running the gadget, we need to get the gadget info to perform
+			// different tasks like creating the parser and setting flags for the
+			// gadget's parameters.
+			actualArgs := cmd.Flags().Args()
```

Comment:
Not related to this PR, but I think dynamic flags for gadget parameters are a mistake. I don't like that gadget parameters like `--ignore-failed` are mixed up with operators/localmanager flags like `--host`  and with generic flags like `--insecure`.

```
$ sudo -E ./ig run trace_exec -h
...
  -h, --help                   help for run
      --host                   Show data from both the host and containers
      --ignore-failed          Ignore failed events (default true)
      --insecure               Allow connections to HTTP only registries
```

I would prefer if it was `--param ignore-failed=true`. It would make it a lot easier for the user to understand the category of the flag. And I suspect the implementation would be easier without the `utils.ParseEarlyFlags`.

Also, I'd like to have an inspect command that would show, among other things, the parameters:
```
$ sudo -E ./ig image inspect trace_exec
```

In this way, the `--help` command would never try to connect to internet.

Reply:
Agree with the `--param` issue and also with having an `inspect` call to show all available information (from the accumulated gadget metadata).

> In this way, the --help command would never try to connect to internet.

Why would it change behavior here? (I mean: which requests would be saved here?)

Reply:
> Why would it change behavior here? (I mean: which requests would be saved here?)

I don't understand the question.

What I meant is that when the user does not really understand how the CLI is supposed to be used and adds a `--help`, it does not print any help message and tries to download something from the internet. I find that confusing for users.

```
$ sudo -E ./ig run trace exec --help
INFO[0000] Experimental features enabled                
Error: getting gadget info: getting gadget image: pulling image "trace": downloading to local repository: failed to resolve ghcr.io/inspektor-gadget/gadget/trace:latest: ghcr.io/inspektor-gadget/gadget/trace:latest: not found
```

Reply:
> Or would you want those params also to move to the more generic solution?

Yes, I would like that. Something like `--operator-param foo=bar`, so the CLI flag `--operator-param` is known without downloading anything. And the user could discover the possible operator params with the `image inspect` command.

Reply:
Hmm, if we go that route, we could also use our internal conventions (e.g. `--param operators.localmanager.foo=abc`). I'm certainly not against it, but it would make things a bit less straightforward than it was before for users.

Another thing I like about it, is that it can differentiate between local and remote operators; the `cli` operator for example is always a local running one, while most of the others are run on the server side. So `cli` could still register params directly, while remote params could be used through the detour. Hmmm.....